pkg/controllers/machinesetmigration/machineset_migration_controller.go

```diff
@@ -0,0 +1,391 @@
+			case corev1.ConditionTrue:
+				isPaused = true
+			}
+		}
+	}
+
+	return isPaused, nil
+}
+
+func (r *MachineSetMigrationReconciler) ensureUnpausedRequestedOnNewAuthoritativeAPI(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet) error {
+	// Request that the new authoritative resource reconciliation is un-paused.
+	switch mapiMachineSet.Spec.AuthoritativeAPI {
+	case machinev1beta1.MachineAuthorityClusterAPI:
+		// For requesting unpausing of a CAPI resource, remove the paused annotation on it.
+		// So check if the ClusterAPI resource has the paused annotation and if so remove it.
+		capiMachineSet := &capiv1beta1.MachineSet{}
+		if err := r.Get(ctx, client.ObjectKey{Namespace: consts.DefaultManagedNamespace, Name: mapiMachineSet.Name}, capiMachineSet); err != nil {
+			return fmt.Errorf("failed to get CAPI machine set: %w", err)
+		}
+
+		if annotations.HasPaused(capiMachineSet) {
+			capiMachineSetCopy := capiMachineSet.DeepCopy()
+			delete(capiMachineSet.Annotations, capiv1beta1.PausedAnnotation)
+
+			if err := r.Patch(ctx, capiMachineSet, client.MergeFrom(capiMachineSetCopy)); err != nil {
+				return fmt.Errorf("failed to patch CAPI machine set: %w", err)
+			}
+		}
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		// For requesting unpausing of a MAPI resource, it is sufficient to switch the spec.AuthoritativeAPI field on the MAPI resource.
+		// which is already done before this code runs in this controller. Nothing to do here.
+	default:
+		// Any other value is disallowed by the openAPI schema validation.
+	}
+
+	return nil
+}
+
+func (r *MachineSetMigrationReconciler) ensurePausedRequestedOnOldAuthoritativeAPI(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet) error {
+	// Request that the old authoritative resource reconciliation is paused.
+	switch mapiMachineSet.Spec.AuthoritativeAPI {
+	case machinev1beta1.MachineAuthorityClusterAPI:
+		// For requesting pausing of a MAPI resource, it is sufficient to switch the spec.AuthoritativeAPI field on the MAPI resource.
+		// which is already done before this code runs in this controller.
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		// For requesting pausing of a CAPI resource, set the paused annotation on it.
+		// The spec.AuthoritativeAPI is set to MachineAPI, meaning that the old authoritativeAPI was ClusterAPI.
+		// So Check if the ClusterAPI resource has the paused annotation, otherwise set it.
+		capiMachineSet := &capiv1beta1.MachineSet{}
+		if err := r.Get(ctx, client.ObjectKey{Namespace: consts.DefaultManagedNamespace, Name: mapiMachineSet.Name}, capiMachineSet); err != nil {
+			return fmt.Errorf("failed to get CAPI machine set: %w", err)
+		}
+
+		if !annotations.HasPaused(capiMachineSet) {
+			capiMachineSetCopy := capiMachineSet.DeepCopy()
+			annotations.AddAnnotations(capiMachineSet, map[string]string{capiv1beta1.PausedAnnotation: ""})
+			if err := r.Patch(ctx, capiMachineSet, client.MergeFrom(capiMachineSetCopy)); err != nil {
```

Comment:
Do we need to requeue after making a change like this?

Reply:
Yes, that'd be better, added.